fxa-rust-client/src/lib.rs

```diff
@@ -145,36 +174,198 @@ impl FirefoxAccount {
+    }
+
+    fn scope_key_contains_scopes(scope_key: &str, scopes: &[&str]) -> bool {
+        for scope in scopes {
+            if !scope_key.contains(scope) {
+                return false;
+            }
         }
-        let client = FxAClient::new(&self.state.config);
-        let session_token = match FirefoxAccount::session_token_from_state(&self.state.login_state)
+        return true;
+    }
+
+    fn oauth_cache_find(&self, scopes: &[&str]) -> Option<&OAuthInfo> {
+        let scope_key = scopes.join("|");
+        // First we try to get the exact same scope.
+        if let Some(info) = self.state.oauth_cache.get(&scope_key) {
+            return Some(info);
+        }
+        // If we can't find it, we look into subsets and TODO: scope implications.
+        for (scope_key, info) in self.state.oauth_cache.iter() {
+            if FirefoxAccount::scope_key_contains_scopes(scope_key, scopes) {
+                return Some(info);
+            }
+        }
+        None
+    }
+
+    pub fn get_oauth_token(&mut self, scopes: &[&str]) -> Result<Option<OAuthInfo>> {
+        let mut refresh_token = None;
+        if let Some(cached_oauth_info) = self.oauth_cache_find(scopes) {
+            if cached_oauth_info.expires_at > util::now_secs() + 60 {
+                return Ok(Some(cached_oauth_info.clone()));
+            }
+            refresh_token = cached_oauth_info.refresh_token.clone();
+        }
+        // This is a bit awkward, borrow checker weirdness.
+        let resp;
+        {
+            if let Some(refresh_token) = refresh_token {
+                let client = FxAClient::new(&self.state.config);
+                resp = client.oauth_token_refresh(OAUTH_CLIENT_ID, &refresh_token, &scopes)?;
+            } else if let Some(session_token) =
+                FirefoxAccount::session_token_from_state(&self.state.login_state)
+            {
+                let client = FxAClient::new(&self.state.config);
+                resp = client.oauth_authorization_with_assertion(
+                    OAUTH_CLIENT_ID,
+                    session_token,
+                    &scopes,
+                )?;
+            } else {
+                return Ok(None);
+            }
+        }
+        Ok(Some(self.handle_oauth_token_response(resp, None)?))
+    }
+
+    pub fn begin_oauth_flow(
+        &mut self,
+        redirect_uri: &str,
+        scopes: &[&str],
+        wants_keys: bool,
+    ) -> Result<OAuthFlowInit> {
+        let flow_id = Uuid::new_v4().simple().to_string();
+        // Create URI
+        let state = FirefoxAccount::random_base64_url_string(16);
+        let code_verifier = FirefoxAccount::random_base64_url_string(43);
+        let code_challenge = hash(MessageDigest::sha256(), &code_verifier.as_bytes())?;
+        let code_challenge = base64::encode_config(&code_challenge, base64::URL_SAFE_NO_PAD);
+        let mut url = Url::parse(&self.state.config.oauth_url)?.join("authorization")?;
```

Comment:
Let's watch out for edge-cases when initiating the flow here, such as https://bugzilla.mozilla.org/show_bug.cgi?id=1456351